### PR TITLE
Use the same semi-infinite scrolling for the notification overview

### PR DESCRIPTION
**Pull request by sukhwinder33445**

### CODE REVIEW

**Review by nilmerg — CHANGES_REQUESTED** [review]
Since this is mostly the same code as for the history overview, I'd like to see a trait being used for this functionality:

* Create a new trait `Icinga\Module\Icingadb\Common\LoadMore`
* Put the new properties of `NotificationList` and methods into this trait (except `init`)
* Use the trait inside `NotificationList` and `HistoryList`
* Expect a `ResultSet` in method `getIterator` as param (that's essentially `realData`)
* Introduce a new method `setLoadMoreUrl` into the trait
  * Use that method in the controller when creating the `NotificationList` and pass it the url with the before param
* Don't add the `ShowMore` widget in the controller, but let `getIterator` add it once the actual data is exhausted. With the url, page size and page number it should be possible to do that.

File: library/Icingadb/Common/LoadMore.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+     */
+    public function setPageNumber($number)
+    {
+        $this->pageNumber = $number;
+
+        return $this;
+    }
+
+    /**
+     * Set the page number
```
Comment (nilmerg) [inline library/Icingadb/Common/LoadMore.php:51]: That's not a page number.

File: library/Icingadb/Common/LoadMore.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+     * @return $this
+     */
+    public function setLoadMoreUrl(Url $url)
+    {
+        $this->loadMoreUrl = $url;
+
+        return $this;
+    }
+
+    protected function getIterator($result)
```
Comment (nilmerg) [inline library/Icingadb/Common/LoadMore.php:64]: * Add phpdoc
* Require that `$result` is a `ResultSet`
  * Then override `__construct` in `HistoryList` and `NotificationList` to also require there that `$data` is a `ResultSet` (but then just call the parent constructor, it's just about the type declaration)

**Review by nilmerg — CHANGES_REQUESTED** [review]
Nice! :+1: Just a few small changes left :)

File: library/Icingadb/Common/LoadMore.php [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+     */
+    public function setPageNumber($number)
+    {
+        $this->pageNumber = $number;
+
+        return $this;
+    }
+
+    /**
+     * Set the url for LoadMore widget
```
Comment (nilmerg) [inline library/Icingadb/Common/LoadMore.php:52]: `LoadMore` doesn't exist as widget, at least I read this as if there's a `Icingadb\Widget\LoadMore`. Just describe what the url is used for: to fetch *more* items

File: library/Icingadb/Common/LoadMore.php (hunk 2 above)
Comment (nilmerg) [reply library/Icingadb/Common/LoadMore.php:70]: Also add the type declaration for the parameter in the method interface.

File: library/Icingadb/Common/LoadMore.php [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+     */
+    public function setLoadMoreUrl(Url $url)
+    {
+        $this->loadMoreUrl = $url;
+
+        return $this;
+    }
+
+    /**
+     * Add the page separator and the "LoadMore" button at the desired position
```
Comment (nilmerg) [inline library/Icingadb/Common/LoadMore.php:66]: That's the description of what it additionally does (if it begins with *Adds*), not the main functionality. Keep that with *Adds* as extended description and add a main description what the method returns. And on that note, also add a `@returns Generator` annotation.

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Review by nilmerg — CHANGES_REQUESTED** [review]
Please check all other usages of `HistoryList` and `NotificationList`. It's not just the EventController which is affected.

**Comment by sukhwinder33445** [comment]
I have already checked. NotificationList is not used anywhere else and HistoryList already gets a Resultset parameter (used in the HostController, ServiceController and in the EventController (fixed)).

**Comment by nilmerg** [comment]
Aye, but it's not just the constructor you've changed :wink:

**Comment by sukhwinder33445** [comment]
Yes, I also changed the getIterator method in the LoadMore trait, but it has no effect. Am I missing something?

**Comment by nilmerg** [comment]
Just an example: https://github.com/Icinga/icingadb-web/blob/feature/add-loadmore-to-notificationsController/application/controllers/HostController.php#L163-L164

**Comment by sukhwinder33445** [comment]
Oh, I see. It is no longer necessary to use ShowMore in the controller. I will change it soon.

**Review by nilmerg — APPROVED** [review]

**Pull request merged**